File: Kakao/Level2/42888-오픈채팅방.js
let STACK = [];
let RELATION = {};

const solution = (record) => {

  for(let i = 0; i < record.length; i++) {
    let stm_split = record[i].split(' ');
    let uuid = stm_split[1];
    if(stm_split.length === 2) {
      leave(uuid);
      continue;
    }
    (stm_split[0] === 'Enter') ? enter(uuid, stm_split[2]) : change(uuid, stm_split[2]);
  }

  let answer = new Array(STACK.length);
  for(let i = 0; i < STACK.length; i++){
    let action = STACK[i][0];
    let uuid = STACK[i][1];
    let name = RELATION[uuid];
    let str = name+"님이 ";
    action === 'ENTER' ? str += "들어왔습니다." : str += "나갔습니다.";
    answer[i] = str;
  }

  return answer;
}

const enter = (uuid, name) => {
  RELATION[uuid] = name;
  STACK.push(['ENTER', uuid]);
}

const leave = (uuid) => {
  STACK.push(['LEAVE', uuid]);
}

const change = (uuid, name) => {
  RELATION[uuid] = name;
}

File: Kakao/Level2/42888-오픈채팅방.py
def solution(record):
  stack = []
  relation = {}
  
  for stm in record:
    split_stm = stm.split(' ')
    if len(split_stm) == 2: # Leave
      leave(split_stm[1], stack)
    else: # Enter or Change
      (action, uuid, name) = split_stm
      if action == 'Enter':
        enter(uuid, name, stack, relation)
      if action == 'Change':
        change(uuid, name, relation)
  
  answer = create_answer(stack, relation)
  return answer
  
  
def enter(uuid, name, stack, relation):
  relation[uuid] = name
  stack.append(('Enter', uuid))

def leave(uuid, stack):
  stack.append(('Leave', uuid))

def change(uuid, new_name, relation):
  relation[uuid] = new_name
  
def create_answer(stack, relation):
  answer = []
  for stm in stack:
    (action, uuid) = stm
    name = relation[uuid]
    ans_stm = name+'님이 '
    if action == 'Enter':
      ans_stm += '들어왔습니다.'
    if action == 'Leave':
      ans_stm += '나갔습니다.'
    answer.append(ans_stm)
  return answer
